gui/services/forms.py

```diff
@@ -439,30 +439,35 @@ def save(self):
         return obj
 
 
-class RsyncdForm(ModelForm):
+class RsyncdForm(MiddlewareModelForm, ModelForm):
+
+    middleware_attr_prefix = "rsyncd_"
+    middleware_attr_schema = "rsyncd"
+    middleware_plugin = "rsyncd"
 
     class Meta:
         fields = '__all__'
         model = models.Rsyncd
 
-    def save(self):
-        super(RsyncdForm, self).save()
-        started = notifier().reload("rsync")
-        if (
-            started is False and
-            models.services.objects.get(srv_service='rsync').srv_enable
-        ):
-            raise ServiceFailed(
-                "rsync", _("The Rsync service failed to reload.")
-            )
-
 
 class RsyncModForm(ModelForm):
 
     class Meta:
         fields = '__all__'
         model = models.RsyncMod
 
+    def save(self):
+        with client as c:
+            cdata = self.cleaned_data
+
+            if self.instance.id:
+                c.call('rsyncdmod.do_update', self.instance.id, cdata)
+            else:
+                c.call('rsyncdmod.do_create', cdata)
+            c.call('service.reload', 'rsync')
+
+        return self.instance
+
     def clean_rsyncmod_name(self):
```

Comment:
Please also move all `clean_*` methods to middleware for both classes. Otherwise their validation and sanitization will be lost in new GUI.

Reply:
@themylogin I've actually haven't moved them to the middlewared intentionally. I have mixed feelings about clean_rsyncmod_name(I'll move it to the middlewared)  but when it comes to the clean_rsyncmod_hostsallow and clean_rsyncmod_hostsdeny methods they are basically removing newline characters. I feel like this is normal data preparation that should be done on the GUI side and middleware should receive payload in more json compliant/standard way. If we move those methods to the middlewared we will have to force new GUI to send newline characters in JSON payload or make this uncessesry ugly and complicated in the middleware plugin. Please correct me if I'm wrong.

Reply:
@650elx you are right, we should use lists there. Look at how `SmartService` handles `email` (a list of e-mail addresses). It is a bit of a hack now (database still stores string and legacy GUI reads directly from database), once we get rid of legacy GUI completely, we'll be able to store them as real lists (probably JSON).